lib/hanami/middleware/body_parser/class_interface.rb

```diff
@@ -0,0 +1,52 @@
+require_relative 'errors'
+require_relative 'parser'
+
+module Hanami
+  module Middleware
+    class BodyParser
+      module ClassInterface
+        # @since x.x.x
+        # @api private
+        def for(parser)
+          if parser_name?(parser)
+            require_parser(parser)
+          elsif parser_class?(parser)
+            parser.new
+          elsif parser_instance?(parser)
+            parser
+          else
+            raise UnknownParserError.new(parser)
+          end
+        end
+
+        private
+
+        # @since 1.3.0
+        # @api private
+        def require_parser(parser)
+          require "hanami/middleware/body_parser/#{parser}_parser"
+
+          parser = Utils::String.classify(parser)
+          Utils::Class.load!("Hanami::Middleware::BodyParser::#{ parser }Parser").new
+        rescue LoadError, NameError
+          raise UnknownParserError.new(parser)
+        end
+
+        # @api private
+        def parser_name?(parser)
+          parser.is_a?(String) || parser.is_a?(Symbol)
+        end
+
+        # @api private
+        def parser_class?(parser)
+          parser.is_a?(Class) && parser < Parser
+        end
+
+        # @api private
+        def parser_instance?(parser)
+          parser.class < Parser
+        end
```

Comment:
I do have some other thoughts about this. Our `BodyParser::Parser` class is really not doing all that much now.

Another approach could be to get rid of it entirely and just add a check here to ensure parser classes conform to a certain API. Something like this:

```ruby
def for(parser)
  parser = 
    if parser_name?(parser)
      require_parser(parser)
    elsif parser.is_a?(Class)
      parser.new
    else
      parser
    end

  ensure_parser parser
  
  parser
end

private

def ensure_parser(parser)
  unless %i[mime_types parse].all? { |method| parser.respond_to?(method) }
    raise UnknownParserError.new(parser)
  end
end
```

This would allow us to get rid of this now pretty useless base class and just use a closer-to-duck-typing kind of check to ensure we have a parser object we can use. The exception name isn't _perfect_ for this kind of case anymore (since it's more of an _invalid_ parser than an _unknown_ parser, but I think we can't change the exception name so easily due to semver commitments?)

Anyway, I feel like this would be the better approach, tbh. It would make things more flexible and allow us to get rid of some unnecessary code.

Would you be happy for me to make that change?

Reply:
What the hey, I'm just going to go ahead and make this change now, since codecov is telling me I don't have test coverage of `Parser`.